File: impls/java/src/mal/step0_repl.java
package mal;

public class step0_repl {

    public String read(String input) {
	return input;
    }
  
    public String eval(String input) {
	return input;
    }
  
    public String print(String input) {
	return input;
    }
  
    public String rep(String input) {
	String afterRead = this.read(input);
	String afterEval = this.eval(afterRead);
	String afterPrint = this.print(afterEval);
	return afterPrint;
    }
  
    public static void main(String... args) {
	String input;
	do {
	    input = System.console().readLine("user> ");
	    if (input != null) {
		System.out.println(input);
	    }
	} while (input != null);
    }
}

File: impls/clojure/src/mal/step0_repl.clj
(ns mal.step0_repl)

(def READ identity)
(def EVAL identity)
(def PRINT identity)

(def rep (comp PRINT EVAL READ))

(defn prompt []
  (print "user> ")
  (flush)
  (read-line))

;; MAIN LOOP

(defn -main [& args]
  (loop [input (prompt)]
    (when input
      (println (rep input))
      (recur (prompt)))))

